go/worker/common/p2p/rpc/client.go

```diff
@@ -0,0 +1,320 @@
+				"peer", peer,
+			)
+
+			rsp := reflect.New(reflect.TypeOf(rspTyp)).Interface()
+			pf, err := c.call(ctx, peer, &request, rsp, maxPeerResponseTime)
+			ch <- &result{rsp, pf, err}
+			close(ch)
+		})
+	}
+
+	// Gather results.
+	var (
+		rsps []interface{}
+		pfs  []PeerFeedback
+	)
+	for _, ch := range resultCh {
+		select {
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		case result := <-ch:
+			// Ignore failed results.
+			if result.err != nil {
+				continue
+			}
+
+			rsps = append(rsps, result.rsp)
+			pfs = append(pfs, result.pf)
+		}
+	}
+	c.logger.Debug("gathered results", "num1", len(rsps), "num2", len(pfs))
+	return rsps, pfs, nil
+}
+
+func (c *client) call(
+	ctx context.Context,
+	peerID core.PeerID,
+	request *Request,
+	rsp interface{},
+	maxPeerResponseTime time.Duration,
+) (PeerFeedback, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	startTime := time.Now()
+
+	err := c.sendRequestAndDecodeResponse(ctx, peerID, request, rsp, maxPeerResponseTime)
+	if err != nil {
+		c.logger.Debug("failed to call method",
+			"err", err,
+			"method", request.Method,
+			"peer_id", peerID,
+		)
+
+		c.RecordFailure(peerID, time.Since(startTime))
+		return nil, err
+	}
+
+	pf := &peerFeedback{
+		mgr:     c.PeerManager,
+		peerID:  peerID,
+		latency: time.Since(startTime),
+	}
+	return pf, nil
+}
+
+func (c *client) sendRequestAndDecodeResponse(
+	ctx context.Context,
+	peerID core.PeerID,
+	request *Request,
+	rsp interface{},
+	maxPeerResponseTime time.Duration,
+) error {
+	// Attempt to open stream to the given peer.
+	stream, err := c.host.NewStream(
+		network.WithNoDial(ctx, "should already have connection"),
+		peerID,
+		c.protocolID,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to open stream: %w", err)
+	}
+	defer stream.Close()
+
+	codec := cbor.NewMessageCodec(stream, codecModuleName)
+
+	// Send request.
+	_ = stream.SetWriteDeadline(time.Now().Add(RequestWriteDeadline))
+	if err = codec.Write(request); err != nil {
+		c.logger.Debug("failed to send request",
+			"err", err,
+			"peer_id", peerID,
+		)
+		return fmt.Errorf("failed to send request: %w", err)
+	}
+	_ = stream.SetWriteDeadline(time.Time{})
+
+	// Read response.
+	// TODO: Add required minimum speed.
+	var rawRsp Response
+	_ = stream.SetReadDeadline(time.Now().Add(maxPeerResponseTime))
+	if err = codec.Read(&rawRsp); err != nil {
+		c.logger.Debug("failed to read response",
+			"err", err,
+			"peer_id", peerID,
+		)
+		return fmt.Errorf("failed to read response: %w", err)
+	}
+	_ = stream.SetWriteDeadline(time.Time{})
+
+	// Decode response.
+	if rawRsp.Error != nil {
+		return errors.FromCode(rawRsp.Error.Module, rawRsp.Error.Code, rawRsp.Error.Message)
+	}
+
+	if rsp != nil {
+		return cbor.Unmarshal(rawRsp.Ok, rsp)
+	}
+	return nil
+}
+
+// NewClient creates a new RPC client for the given protocol.
+func NewClient(p2p P2P, runtimeID common.Namespace, protocolID, version string) Client {
```

Comment:
Change version type to `version.Version` (and also on the server) and in `NewRuntimeProtocolID` could then either use the full version or the major part only.